src/components/settings/PushNotifications/logic.ts

```diff
@@ -0,0 +1,123 @@
+import { keccak256, toUtf8Bytes } from 'ethers/lib/utils'
+import { getToken, getMessaging } from 'firebase/messaging'
+import { DeviceType } from '@safe-global/safe-gateway-typescript-sdk'
+import type { RegisterNotificationsRequest } from '@safe-global/safe-gateway-typescript-sdk'
+import type { Web3Provider } from '@ethersproject/providers'
+
+import { FIREBASE_VAPID_KEY, initializeFirebaseApp } from '@/services/push-notifications/firebase'
+import { trackEvent } from '@/services/analytics'
+import { PUSH_NOTIFICATION_EVENTS } from '@/services/analytics/events/push-notifications'
+import packageJson from '../../../../package.json'
+import { logError } from '@/services/exceptions'
+import ErrorCodes from '@/services/exceptions/ErrorCodes'
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
+
+// We store UUID locally to track device registration
+export type NotificationRegistration = WithRequired<RegisterNotificationsRequest, 'uuid'>
+
+export const requestNotificationPermission = async (): Promise<boolean> => {
+  if (Notification.permission === 'granted') {
+    return true
+  }
+
+  let permission: NotificationPermission | undefined
+
+  try {
+    permission = await Notification.requestPermission()
+  } catch (e) {
+    logError(ErrorCodes._400, e)
+  }
+
+  const isGranted = permission === 'granted'
+
+  trackEvent(isGranted ? PUSH_NOTIFICATION_EVENTS.GRANT_PERMISSION : PUSH_NOTIFICATION_EVENTS.REJECT_PERMISSION)
+
+  if (!isGranted) {
+    alert('You must allow notifications to register your device.')
+  }
```

Comment:
We show this alert on the global preferences but not when on a specific safe. Is there a reason for that?

Reply:
I've removed this. The function is now being used as an early return and I've added a comment explaining it in 12116b71f0067a697e64c861fc682219a9014108.